### PR TITLE
Introduced `RandomScheduler`

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

 Same as `MixedScheduler`, but performs tasks with the same priority randomly
 
 It is useful in scenarios where you want to introduce non-deterministic behavior in task execution while still
 ensuring that higher-priority tasks are generally executed before lower-priority ones.
 Uses a Pseudo-Random Number Generator (PRNG) for selecting the next task to execute within a tasks with the same priority.
 `RandomScheduler` can be seeded for reproducibility.

## Related

- Related Issue #
- Closes #

### CODE REVIEW

File: .changeset/cyan-coats-peel.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"effect": patch
```
Comment (tim-smart) [inline .changeset/cyan-coats-peel.md:2]:
```suggestion
"effect": minor
```

File: packages/effect/src/Scheduler.ts [hunk 2]
```diff
@@ -238,6 +240,102 @@ export class ControlledScheduler implements Scheduler {
...
+/**
+ * Same as `MixedScheduler`, but performs tasks with the same priority randomly
+ *
+ * It is useful in scenarios where you want to introduce non-deterministic behavior in task execution while still
+ * ensuring that higher-priority tasks are generally executed before lower-priority ones.
+ * Uses a Pseudo-Random Number Generator (PRNG) for selecting the next task to execute within a tasks with the same priority.
+ * `RandomScheduler` can be seeded for reproducibility.
+ *
+ * @see MixedScheduler
+ * @since 3.6.4
```
Comment (tim-smart) [inline packages/effect/src/Scheduler.ts:252]:
```suggestion
 * @since 3.7.0
```

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: .changeset/cyan-coats-peel.md (hunk 1 above)
Comment (KhraksMamtsov) [reply .changeset/cyan-coats-peel.md]: done

File: packages/effect/src/Scheduler.ts (hunk 2 above)
Comment (KhraksMamtsov) [reply packages/effect/src/Scheduler.ts]: done

**Comment by mikearnaldi** [comment]
Order of execution within the same priority is semantically important, introducing non determinism at that level may lead to very hard to debug issues and is unwarranted. I think if you want to introduce non-determinism you can do so by yielding to a random priority instead of processing with a random order

**Comment by KhraksMamtsov** [comment]
```
[1,[a,b,c]]
[2,[d,e,f]]

[c-b-a, e-d-f] // respect priority but not order inside priority
[d-e-f, a-b-c] // respect order inside priority but no priority itself
```
in the first case, the semantics of priorities are preserved.
and in the second case, the priorities can simply be called a group of tasks.
Are you sure about this?

**Comment by mikearnaldi** [comment]
```
I believe that usage RandomScheduler will highlight this kind of bugs.
And we shouldn't rely on execution order within one priority and use correct concurrent primitives correctly)
It is assumed that it will be used in testing
```

No, order of execution has to be guaranteed otherwise semantics are broken. That's a required invariant for Effect to work, we have to rely on order of execution.

**Comment by KhraksMamtsov** [comment]
Ok, It's clear, thank you

**Comment by KhraksMamtsov** [comment]
> No, order of execution has to be guaranteed otherwise semantics are broken. That's a required invariant for Effect to work, we have to rely on order of execution. Both priority and order within priority.

I think it should be reflected somewhere in the documentation.

**Comment by KhraksMamtsov** [comment]
Also I am glad that another bug was discovered and fixed during the development of this PR)

**Comment by mikearnaldi** [comment]
Which bug? Re docs we can add it to the scheduler docs, it's fairly low level to make schedulers manually and there can be a degree of freedom e.g. one may decide to execute random within a specific priority that is not used elsewhere. Will try to find a good explainer

**Comment by KhraksMamtsov** [comment]
@mikearnaldi 
https://discord.com/channels/795981131316985866/795983589644304396/1273026396779974807
https://github.com/Effect-TS/effect/pull/3452

**Comment by mikearnaldi** [comment]
Ah yeah indeed

**Pull request closed without merging**